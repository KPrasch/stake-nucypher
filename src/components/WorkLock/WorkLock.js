import React, { Component } from 'react';
import { Form, Col, Row, Button } from 'react-bootstrap';
import { Formik } from 'formik';
import * as Yup from 'yup';

const validationSchema = Yup.object().shape({
  bidValue: Yup.string().matches(/[0-9+]/).required('Field is required.')
});

function WorkLock(props) {
  const onSubmit = (value) => {
    if (props.onBid) {
      props.onBid(value);
    }
  }
  return <Formik onSubmit={onSubmit} initialValues={{ bidValue: 0 }} validationSchema={validationSchema}>
    {({
      handleSubmit,
      handleChange,
      handleBlur,
      values,
      touched,
      isValid,
      errors,
      setValues,
    }) => {
        const onEthValueChange = (...args) => {
          props.onEthValueChange && props.onEthValueChange(...args);
          handleChange(...args);
        };
        return <>
          <Form onSubmit={handleSubmit} className="w-100">
            <Form.Group as={Row}>
              <Form.Label column md={{ span: 3, offset: 1 }} htmlFor="bid-value">
                ETH amount
              </Form.Label>
              <Col md={6}>
                <Form.Control id="bid-value" size="bg" type="text" name="bidValue" onChange={onEthValueChange} value={values.bidValue} />
                <Form.Control.Feedback type="invalid" style={{ display: errors.bidValue ? 'inline' : 'none' }}>{errors.bidValue}</Form.Control.Feedback>
              </Col>
            </Form.Group>
            <div className="d-flex justify-content-center">
              <Button type="submit" disabled={!isValid}>Bid</Button>
            </div>
          </Form>
        </>;
      }
    }
    </Formik>;
}

export default WorkLock;
