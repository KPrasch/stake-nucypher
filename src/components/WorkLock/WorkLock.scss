
.feedback-placeholder {
  height: 20px;
}
