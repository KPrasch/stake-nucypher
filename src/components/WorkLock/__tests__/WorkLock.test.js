import React from 'react';
import WorkLock from '../WorkLock';
import { render, fireEvent, wait } from '@testing-library/react'

describe('<WorkLock />', () => {

  it('should render correctly', () => {
    const { asFragment } = render(<WorkLock />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should trigger onBid on bid', async () => {
    const onBid = jest.fn(() => {});
    const { getByLabelText, getByText } = render(<WorkLock onBid={onBid} />);

    await wait(() => {
      fireEvent.change(getByLabelText('ETH amount'), { target: { value: '6' } });
      fireEvent.click(getByText('Bid'));
    });

    expect(onBid).toBeCalledWith(expect.objectContaining({ bidValue: '6' }));
  });

  it('should validate input from user correctly', async () => {
    const { getByLabelText, getByText, asFragment } = render(<WorkLock />);

    await wait(() => {
      fireEvent.change(getByLabelText('ETH amount'), { target: { value: '' } });
      fireEvent.click(getByText('Bid'));
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
