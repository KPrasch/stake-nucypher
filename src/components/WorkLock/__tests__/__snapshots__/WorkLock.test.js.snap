// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WorkLock /> should render correctly 1`] = `
<DocumentFragment>
  <form
    class="w-100"
  >
    <div
      class="form-group row"
    >
      <label
        class="form-label col-form-label col-md-3 offset-md-1"
        for="bid-value"
      >
        ETH amount
      </label>
      <div
        class="col-md-6"
      >
        <input
          class="form-control form-control-bg"
          id="bid-value"
          name="bidValue"
          type="text"
          value="0"
        />
        <div
          class="invalid-feedback"
          style="display: none;"
        />
      </div>
    </div>
    <div
      class="d-flex justify-content-center"
    >
      <button
        class="btn btn-primary"
        type="submit"
      >
        Bid
      </button>
    </div>
  </form>
</DocumentFragment>
`;

exports[`<WorkLock /> should validate input from user correctly 1`] = `
<DocumentFragment>
  <form
    class="w-100"
  >
    <div
      class="form-group row"
    >
      <label
        class="form-label col-form-label col-md-3 offset-md-1"
        for="bid-value"
      >
        ETH amount
      </label>
      <div
        class="col-md-6"
      >
        <input
          class="form-control form-control-bg"
          id="bid-value"
          name="bidValue"
          type="text"
          value=""
        />
        <div
          class="invalid-feedback"
          style="display: inline;"
        >
          Field is required.
        </div>
      </div>
    </div>
    <div
      class="d-flex justify-content-center"
    >
      <button
        class="btn btn-primary"
        disabled=""
        type="submit"
      >
        Bid
      </button>
    </div>
  </form>
</DocumentFragment>
`;
